clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -730,8 +730,15 @@ private void close(long timeout, TimeUnit timeUnit, boolean swallowException) {
         ClientUtils.closeQuietly(partitioner, "producer partitioner", firstException);
         AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId);
         log.debug("The Kafka producer has closed.");
-        if (firstException.get() != null && !swallowException)
-            throw new KafkaException("Failed to close kafka producer", firstException.get());
+        Throwable exception = firstException.get();
+        if (exception != null && !swallowException) {
+            if (exception instanceof InterruptedException) {
+                throw new InterruptException("kafka producer thread was interrupted");
+            } else if (exception instanceof InterruptException) {
```

Comment:
@ijuma , sorry for the delay. I am not sure actually. Just the defensive programming. Do you think it should be refined?